### PR TITLE
Release UI: detect devmode revisions

### LINKED ISSUE #1186: Dev mode/grade detection to disable stable/candidate channels

### PR DESCRIPTION
**Pull request by bartaz**
Fixes #1186 

Adds information about devmode revisions (if there are any in history or currently released) and prevents from releasing devmode snaps to stable and candidate.

### QA
- ./run or http://snapcraft.io-canonical-websites-pr-1195.run.demo.haus/
- go to Releases page of a snap you owned
- if there are any revisions in history that are devmode they should be marked with an icon
- icon should have a tooltip
- if any devmode/devel revision is released (in edge or beta) warning icon should appear next to it, tooltip should say why the icon is there and notification should appear with links to the docs.

screen shot 2018-10-12 at 16 35 02

### CODE REVIEW

**Comment by bartaz** [comment]
Not sure if current icon is not **warning** too much, but we don't have any other ATM.

File: static/js/publisher/release/revisionsTable.js [hunk 1]
```diff
@@ -156,6 +177,17 @@ export default class RevisionsTable extends Component {
...
 
+        // check for devmode revisions
+        if (risk === EDGE || risk === BETA || risk === UNASSIGNED) {
+          const hasDevmodeRevisions = Object.values(nextChannelReleases[channel]).some(isInDevmode);
+
+          // remove stable and beta channels as targets if any revision
+          // is in devmode
+          if (hasDevmodeRevisions) {
+            targetRisks = targetRisks.slice(2);
+          }
```
[the remark this answers is not in the document]
Comment (bartaz) [reply static/js/publisher/release/revisionsTable.js]: Good point. If possible I'd like to address this separately, as currently Vanilla context menu doesn't seem to support 'disabled' items, so I would need to add that.

File: static/js/publisher/release/releasesController.js [hunk 2]
```diff
@@ -400,13 +401,26 @@ export default class ReleasesController extends Component {
...
         { this.state.error &&
           <Notification status="error" appearance="negative">
             {this.state.error}
           </Notification>
         }
+        { hasDevmodeRevisions &&
+          <Notification appearance="caution">
+            Revisions in development mode cannot be released to stable or candidate channels.
+            <br/>
+            You can read more about <code>devmode</code> confinement and <code>devel</code> grade in <a href="https://docs.snapcraft.io/build-snaps/syntax">the Snapraft docs</a>.
```
[the remark this answers is not in the document]
Comment (bartaz) [reply static/js/publisher/release/releasesController.js]: I had 2 separate links for devmode and devel grade in first place, but then new docs were published and the content was gone. Now I found the new confinement page (https://docs.snapcraft.io/t/snap-confinement/6233), so I linked to that.

**Comment by bartaz** [comment]
@matthewpaulthomas I added a small table column on the left side of version just to contain the icon. Both icons and version strings are always aligned. I think it looks better then before even without any 'non-warning' icon.

**Comment by matthewpaulthomas** [comment]
Ah, clever.

+1 if you do the same with the icons in the “Releases available” table as you’ve done for the “Revisions available” table.

File: static/js/publisher/release/revisionsTable.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (matthewpaulthomas) [reply static/js/publisher/release/revisionsTable.js]: Ok, reported as #1204.

**Pull request merged**